sdk/dynamicconfig/sdk/metric/sdk.go

```diff
@@ -0,0 +1,625 @@
+		shouldCollect := m.shouldCollect(name, periods)
+		if !shouldCollect {
+			return true
+		}
+
+		mods := atomic.LoadInt64(&inuse.updateCount)
+		coll := inuse.collectedCount
+
+		if mods != coll {
+			// Updates happened in this interval,
+			// checkpoint and continue.
+			checkpointed += m.checkpointRecord(inuse)
+			inuse.collectedCount = mods
+			return true
+		}
+
+		// Having no updates since last collection, try to unmap:
+		if unmapped := inuse.refMapped.tryUnmap(); !unmapped {
+			// The record is referenced by a binding, continue.
+			return true
+		}
+
+		// If any other goroutines are now trying to re-insert this
+		// entry in the map, they are busy calling Gosched() awaiting
+		// this deletion:
+		m.current.Delete(inuse.mapkey())
+
+		// There's a potential race between `LoadInt64` and
+		// `tryUnmap` in this function.  Since this is the
+		// last we'll see of this record, checkpoint
+		mods = atomic.LoadInt64(&inuse.updateCount)
+		if mods != coll {
+			checkpointed += m.checkpointRecord(inuse)
+		}
+		return true
+	})
+
+	return checkpointed
+}
+
+// CollectAsync implements internal.AsyncCollector.
+func (m *Accumulator) CollectAsync(kv []kv.KeyValue, obs ...metric.Observation) {
+	labels := label.NewSetWithSortable(kv, &m.asyncSortSlice)
+
+	for _, ob := range obs {
+		if a := m.fromAsync(ob.AsyncImpl()); a != nil {
+			a.observe(ob.Number(), &labels)
+		}
+	}
+}
+
+func (m *Accumulator) observeAsyncInstruments(
+	ctx context.Context,
+	periods []pb.ConfigResponse_MetricConfig_Schedule_CollectionPeriod,
+) int {
+	m.asyncLock.Lock()
+	defer m.asyncLock.Unlock()
+
+	asyncCollected := 0
+
+	// TODO: change this to `ctx` (in a separate PR, with tests)
+	m.asyncInstruments.Run(context.Background(), m)
+
+	for _, inst := range m.asyncInstruments.Instruments() {
+		if a := m.fromAsync(inst); a != nil &&
+		m.shouldCollect(a.descriptor.Name(), periods) {
+			asyncCollected += m.checkpointAsync(a)
+		}
+	}
+
+	return asyncCollected
+}
+
+// Determines if we should collect metrics from an instrument with `name`.
+func (m *Accumulator) shouldCollect(
+	name string,
+	periods []pb.ConfigResponse_MetricConfig_Schedule_CollectionPeriod,
+) bool {
+	// If we aren't enabling the dynamic config extension, we collect everything.
+	if m.ext == nil {
+		return true
+	}
+
+	m.ext.lock.Lock()
+	// Determine CollectionPeriod associated with instrument.
+	var collectionPeriod pb.ConfigResponse_MetricConfig_Schedule_CollectionPeriod = 0
+	if period, ok := m.ext.instrumentPeriod[name]; ok {
+		collectionPeriod = period
+	} else {
+		collectionPeriod = FindPeriod(name, m.ext.schedules)
+		m.ext.instrumentPeriod[name] = collectionPeriod
+	}
+	m.ext.lock.Unlock()
+
+	// We treat a collectionPeriod of 0 as meaning "do not collect".
+	if collectionPeriod == 0 {
+		return false
+	}
+
+	// CollectionPeriod of instrument must match one of the `periods`
+	for _, period := range periods {
+		if collectionPeriod == period {
+			return true
+		}
+	}
```

Comment:
Would it be possible to use all the periods specified in the protobuf, rather than passing in a specific list of periods? Would there be a use case in which a user would like to restrict the set of periods?

Reply:
Ah, sorry about this being unclear.

We pass the `Collect()` method an optional parameters of `periods`. `periods` is a list of all the collection periods whose instruments we should collect from right now. So if we want to collect from instruments with a period of 1 minute and 10 minutes, we call collect with a list of [1_minute 10_minutes].

Reply:
Let's add a comment.